Ensure MaskedColumn(masked_column) propagates info.serialize_method.

mhvk opened the pull request:
There was a work-around for that in io/ascii/ui.py that can now be removed.

Also ensured .view and .copy propagate .info correctly by adjusting `__array_finalize__`.

Despite the latter, felt this was more of a refactoring than a bug fix, but am happy to change that (in which case this would be for 4.3.1, I guess).

Saw this while trying to get `Masked` to propagate properly and wondered if this also was an 'info doesn't propagate right' problem. Turns out it was! Still puzzling a bit over why the table is copied there (#7605), and will see if I can solve that too. But one issue at a time!

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label.
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

mhvk commented:
Oops, this needs the small `data_info` bug fix from #11910; I must have been on that branch while testing this...

taldcroft commented:
This has a merge conflict now and I'm a bit confused about this and #11919.

taldcroft commented on astropy/table/column.py:
```diff
@@ -536,6 +536,8 @@ def __array_finalize__(self, obj):
         if not hasattr(self, 'indices'):  # may have been copied in __new__
             self.indices = []
         self._copy_attrs(obj)
+        if 'info' in getattr(obj, '__dict__', {}):
```
This does not appear to be directly tested. Do you have cases in mind that fail in current main?

mhvk replied:
I'm not sure anymore... But on current master:
```
In [4]: mc = MaskedColumn([1, 2, 3], mask=[True, False, True])

In [5]: mc.info.serialize_method
Out[5]: 
{'fits': 'null_value',
 'ecsv': 'null_value',
 'hdf5': 'data_mask',
 None: 'null_value'}

In [6]: mc.info.serialize_method['fits'] = 'data_mask'

In [7]: mc12 = mc[1:]

In [8]: mc12.info.serialize_method
Out[8]: 
{'fits': 'null_value',
 'ecsv': 'null_value',
 'hdf5': 'data_mask',
 None: 'null_value'}
```

Which suggests I should add a test for this!

mhvk replied:
Actually, the `.view` test does go through this path. But since I couldn't remember that, I added a slice tests too.

mhvk replied:
And the test for the slice brought out that yet another fix was needed... Now done.

taldcroft reviewed (commented):
The `MaskedColumn` changes and tests look straightforward.

mhvk commented:
OK, rebased. Hopefully tests still pass.

The goal here is just to ensure that any initialization, copy, or view of `MaskedColumn` propagates `.info` correctly, including `info.serialize_method`. With that done, I could remove the work-around in `ui.py` for that not working correctly before. But it is not directly related to the goal of #11919, which was to avoid needless copies of the column data. (Of course, the two *are* related in that the problem with `serialize_method` not propagating led me to look at `ui.py`, where I saw the old comment that it wasn't so nice that a copy of the table was needed and thought I should try to understand what happened!)

taldcroft commented on astropy/table/column.py:
```diff
@@ -1393,6 +1395,10 @@ def __new__(cls, data=None, name=None, mask=None, fill_value=None,
         self = ma.MaskedArray.__new__(cls, data=self_data, mask=mask)
+        # The above process preserves info relevant for Column, but this does not
+        # include serialize_method relevant for MaskedColumn.
+        if isinstance(data, MaskedColumn) and 'info' in getattr(data, '__dict__', {}):
```
I haven't thought about the details but why is `serialize_method` being singled out here instead of doing basically `self.info = data.info` to get any attributes that are there now or might be defined in the future? And why doesn't this get done already from the `Column` array finalize?

Sorry for being lazy and not reminding myself how this all works in detail...

mhvk replied:
Mostly because the basecolumn did some of the `info` copying. But, really, you are right! I'll change that and check that it works.

mhvk replied:
Oops, no, for this to work, we need to ensure info attribute names are not assumed to be identical, i.e., we need https://github.com/astropy/astropy/pull/11910/files#diff-72a70f056e1f7a7bad3e2e054326c7b17d2a815d0bf7a31b2bd59df2af039aa3R360

I've pulled it out of #11910 since it helps this PR we are now discussing (and is probably more obviously correct here). I'll rebase #11910 later.

mhvk commented:
Well, your review brought up quite a few good points! I think I've addressed them all...

taldcroft commented:
@mhvk - sorry for the long delay (vacation + work slam). This looks fine now, but I do think it is a bug fix since the demonstration you showed is not the desired behavior. So this just needs a change log and can probably be milestoned for 4.3.1.

mhvk commented:
OK, thanks! I merged #11910 first, rebased this one to remove the duplication, added a changelog entry, and set the milestone for 4.3.1. Tests seem to pass except coverage, which is a red herring.

pllim commented:
I remilestoned to 4.3.2, as 4.3.1 is coming out soon with a hot fix. FYI

mhvk commented:
@taldcroft - I think this is ready to go in!

taldcroft commented on docs/changes/table/11917.bugfix.rst:
```diff
@@ -0,0 +1,3 @@
+Ensured that ``MaskedColumn.info`` is propagated in all cases, so that when
+tables are sliced, writing will still be as reques on
```
Close but no cigar. :smile:  `reques` => `requested`.

mhvk replied:
Darn!  Fixed...

taldcroft reviewed (approved):
Approved once the typo is fixed.
